### ISSUE TITLE
Commit with overflow date field does not raise concerns

### ISSUE THREAD

**Issue by ardumont** [jelmer/dulwich#567]
Hello,

## TL;DR

As a user, i'd like to be notified when a commit field date (author/committer) can cause overflow error.

## Details

I have a repository on disk which holds a well formatted commit except
that the author/committer date can cause overflow error.
But I don't see checks on those fields even though some checks exists for other objects's date field (tag for example).

Here is a partial example commit with such properties:

```
commit 24b03c53f91ec7719928abfeee77d61ad995f5ed
tree 7b6189adcaa2f420f4aa44e1550c3a7de659ad2f
parent 416c46f96c648c0effb9f6adccae3e10d2d7b9f9
author Igor Sysoev <igor@sysoev.ru> 18446743887488505614 +42707004
committer Kirill A. Korinskiy <catap@catap.ru> 18446743887488505614 +42707004

    nginx 0.7.63
    ...
```

Loading it through dulwich, everything seems fine since the object is found:
```
>>> from dulwich.repo import Repo
>>> r = Repo('./mainline.git')
>>> c = r[b'24b03c53f91ec7719928abfeee77d61ad995f5ed']
>>> c
<Commit b'24b03c53f91ec7719928abfeee77d61ad995f5ed'>
>>> c._author_time
18446743887488505614
>>> c.check()
>>>
```

`git fsck` (git version 1:2.11.0-3+deb9u2) reveals that error:

```
git fsck 2>&1 | grep 24b03c53f91ec7719928abfeee77d61ad995f5ed
error in commit 24b03c53f91ec7719928abfeee77d61ad995f5ed: badDateOverflow: invalid author/committer line - date causes integer overflow
```

But, there exists checks on date for other objects (tag for
one). For example, taking another repository with a wrong release date:
```
$ git show --format=raw ae51106031a0bb39a8def57a8592f70116487eab
# git show does not display the date here
#...
tag HYLAFAX-4_4_4-rc1
Tagger: Aidan Van Dyk <aidan@ifax.com>

commit 222d9a22c53ab359cc1b1bf94d01c34369fa2d5c
tree 802f3ede8ded6240e1c9d15ee4a1a29165d1fb8d
parent 815a7290a929509e1a4ffbbc493d25457f10def6
author Aidan Van Dyk <aidan@ifax.com> 1198083348 +0000
committer Aidan Van Dyk <aidan@ifax.com> 1198083348 +0000

    HylaFAX 4.4.4-rc1

diff --git a/VERSION b/VERSION
index 9e3a933..2c78c2d 100644
--- a/VERSION
+++ b/VERSION
@@ -1 +1 @@
-4.4.3
+4.4.4-rc1

$ git fsck 2>&1 | grep ae51106031a0bb39a8def57a8592f70116487eab
error in tag ae51106031a0bb39a8def57a8592f70116487eab: missingSpaceBeforeDate: invalid author/committer line - missing space before date
```

Then trying to have that object with dulwich, this raises an ObjectFormatException which is great:
```
>>> r = Repo('./hylafax.git')
>>> c = r[b'ae51106031a0bb39a8def57a8592f70116487eab']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/dulwich/repo.py", line 472, in __getitem__
    return self.object_store[name]
  File "/usr/lib/python3/dist-packages/dulwich/object_store.py", line 119, in __getitem__
    return ShaFile.from_raw_string(type_num, uncomp, sha=sha)
  File "/usr/lib/python3/dist-packages/dulwich/objects.py", line 393, in from_raw_string
    obj.set_raw_string(string, sha)
  File "/usr/lib/python3/dist-packages/dulwich/objects.py", line 305, in set_raw_string
    self.set_raw_chunks([text], sha)
  File "/usr/lib/python3/dist-packages/dulwich/objects.py", line 310, in set_raw_chunks
    self._deserialize(chunks)
  File "/usr/lib/python3/dist-packages/dulwich/objects.py", line 765, in _deserialize
    raise ObjectFormatException("Unknown field %s" % field)
dulwich.errors.ObjectFormatException: Unknown field b'Thu'
```

What i am aiming at is would it be possible to have the same behavior
for commits?

## Enhancement proposal

Would you be interested to add some check ([function
`dulwich.objects.Commit.check`](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1171-L1175))
on those cases? Or are there reasons for not checking those?

I'm willing to dig in and propose a PR.  The solution i'm thinking of
would be to align with the current behavior:

- check the author/committer date ([in the internal consistency check
function](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1171-L1175))

- if check fails, raise an ObjectFormatException (as explained in that
  function's docstring, `malformed in some way`).

Notes:

- [the commit which initialized the checks](https://github.com/jelmer/dulwich/commit/eb2afb8aeafdaab56fe50d0a14e50d727ebc39f6) aimed at *roughly* reproduce `git fsck`'s
  behavior (according to the  git commit message). That would be a good enhancement.

- [there is comment in that method](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1180)
  which says that the time are tested when set.  I did not see those tests though.

- I'm running debian stretch and currently using version:
```
$ dpkg -l python3-dulwich
ii  python3-dulwich         0.18.4-1         amd64            Python Git library - Python3 module
```

Cheers,

**Comment by ardumont** [comment]
After some more digging, the error as seen in the stacktrace (for the tag with invalid date) is not raised from the check method but from the _deserialize one.

> there is comment in that method which says that the time are tested when set. I did not see those tests though.

This may be what the comment i did not understood earlier is all about.

**Comment by jelmer** [comment]
Hi Antoine,

On Mon, Oct 30, 2017 at 04:26:27AM -0700, Antoine R. Dumont wrote:
 
> ## Enhancement proposal
> 
> Would you be interested to add some check ([function
> `dulwich.objects.Commit.check`](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1171-L1175))
> on those cases? Or are there reasons for not checking those?
> 
> I'm willing to dig in and propose a PR.  The solution i'm thinking of
> would be to align with the current behavior:
> 
> - check the author/committer date ([in the internal consistency check
> function](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1171-L1175))
> 
> - if check fails, raise an ObjectFormatException (as explained in that
>   function's docstring, `malformed in some way`).
> 
> Notes:
> 
> - [the commit which initialized the checks](https://github.com/jelmer/dulwich/commit/eb2afb8aeafdaab56fe50d0a14e50d727ebc39f6) aimed at *roughly* reproduce `git fsck`'s
>   behavior (according to the  git commit message). That would be a good enhancement.
> 
> - [there is comment in that method](https://github.com/jelmer/dulwich/blob/master/dulwich/objects.py#L1180)
>   which says that the time are tested when set.  I did not see those tests though.

It would be great Dulwich was extended to provide similar validation as C git.
Doing this in either the deserialization stage or in check would make sense to
me.

Jelmer

**Comment by ardumont** [comment]
> It would be great Dulwich was extended to provide similar validation as C git. Doing this in either the deserialization stage or in check would make sense to me.

Awesome.
Working on it.

Cheers,
